chatsky/pipeline/service/extra.py

```diff
@@ -35,29 +35,39 @@ class ComponentExtraHandler(BaseModel, extra="forbid", arbitrary_types_allowed=T
     A component extra handler is a set of functions, attached to pipeline component (before or after it).
     Extra handlers should execute supportive tasks (like time or resources measurement, minor data transformations).
     Extra handlers should NOT edit context or pipeline, use services for that purpose instead.
-
-    :param functions: A list or instance of :py:data:`~.ExtraHandlerFunction`.
-    :type functions: :py:data:`~.ExtraHandlerFunction`
-    :param stage: An :py:class:`~.ExtraHandlerType`, specifying whether this handler will be executed before or
-        after pipeline component.
-    :param timeout: (for asynchronous only!) Maximum component execution time (in seconds),
-        if it exceeds this time, it is interrupted.
-    :param requested_async_flag: Requested asynchronous property.
     """
 
     functions: List[ExtraHandlerFunction] = Field(default_factory=list)
+    """
+    A list or instance of :py:data:`~.ExtraHandlerFunction`.
+    """
     stage: ClassVar[ExtraHandlerType] = ExtraHandlerType.UNDEFINED
+    """
+    An :py:class:`~.ExtraHandlerType`, specifying whether this handler will
+    be executed before or after pipeline component.
+    """
     timeout: Optional[float] = None
+    """
+    (for asynchronous only!) Maximum component execution time (in seconds),
+    if it exceeds this time, it is interrupted.
+    """
     requested_async_flag: Optional[bool] = None
+    """
+    Requested asynchronous property.
+    """
 
     @model_validator(mode="before")
     @classmethod
-    # Here Script class has "@validate_call". Is it needed here?
     def functions_constructor(cls, data: Any):
-        if not isinstance(data, dict):
+        if isinstance(data, (list, Callable)):
             result = {"functions": data}
-        else:
+        elif isinstance(data, dict):
             result = data.copy()
+        else:
+            raise ValueError(
+                "Extra Handler can only be initialized from a Dict,"
+                " a Callable or a list of Callables. Wrong inputs received."
+            )
```

Comment:
No, `data` can be anything (e.g. it can be an instance of `ExtraHandler`).
Just do `return data` and pydantic will handle any type mismatches on its own.

So the code would be:
```python
if isinstance(data, list):
  return {"functions": data}
if callable(data):
  return {"functions": [data]}
return data
```

Same with some other validators.

Reply:
Fixed this, but my code can't look exactly like this, because `dict` has to be additionally validated - "functions" have to be cast to a list, since the passed `dict` could be anything.